docs: add proposal of karmadactl init support config file (#5277)
karmada-io/karmada: docs/proposals/karmadactl/Karmadactl Supports Configuration File Method for Production-Grade Installation and Deployment.md

Reviewer: The file name seems too long, maybe we can give it a short one, like `init-configuration-file.md`.

Author: Ok, I have revised the questions here and look forward to your more suggestions
@RainbowMango

---
title: Karmadactl Supports Configuration File Method for Production-Grade Installation and Deployment
authors:

>   - "@tiansuo114"
>   - "@liangyuanpeng"

Reviewer 2: ```suggestion
  - "@tiansuo114"
```
You can put me into reviewer :)

Author: Ok, I've noticed that I might also need to add some specific data structure content, which I'll do in the next revision

### User Stories (Optional)

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### CLI flags changes

This proposal proposes new flags `config`in `karmadactl init` flag set.

| name   | shorthand | default | usage                  |
| ------ | :-------: | ------- | ---------------------- |
| config |     /     | ""      | Karmada init file path |

With these flag, we will:

- When the user has a karmadactl init file, they can use the `--config` flag to specify that `karmadactl init` should read the configuration from the specified config file. Additionally, external cmd flags can still be used. In the design, if there are overlapping parts between the two, the configuration file configuration items will take higher priority.

### Adding New Commands

#### karmadactl config image list

> - description:Shows information about the images required for Karmada deployment.

Author: Yes, I plan to call the ReleaseVersion method in the Karmada library when constructing the output string, so that the image version obtained is consistent with the version required for deployment.

> - description:Shows information about the images required for Karmada deployment.

Reviewer 3: Can it be specified through a parameters?

> - description:Shows information about the images required for Karmada deployment.

Author: At present, there is no design method to specify by parameters, because the deployment requires different image sources, some from kubernets, some from karmada, if you need to specify by parameters version may need to use different parameters, or need further discussion and design

> - description:Shows information about the images required for Karmada deployment.

Reviewer 3: So, what is actually being displayed is the default image tag?

> - description:Shows information about the images required for Karmada deployment.

Author: Yes, it displays the default image tag by default

- usage scope:Default image and configure the specified image
- scope-related flag set:

| name                      | shorthand | default | usage                                                        |
| ------------------------- | --------- | ------- | ------------------------------------------------------------ |
| private-image-registry    | /         | ""      | Private image registry where pull images from. If set, all required images will be downloaded from it, it would be useful in offline installation scenarios. |
| kube-image-mirror-country | /         | ""      | Country code of the kube image registry to be used. For Chinese mainland users, set it to cn |

- performance：

>   1. When the `private-image-registry` and `kube-image-mirror-country` flags are empty, the `list` command prints the images required for deploying Karmada under default settings, using the default image repository `registry.k8s.io`.

Author: In my design, the current default is to display the images of all components in karmada init. I noticed that the karmada-descheduler plugin and other plugins belong to the karmada addons command. I think it would be beneficial to add a new cmd flag to the config image in the future, to display the installation addresses of these plugin images.

>   1. When the `private-image-registry` and `kube-image-mirror-country` flags are empty, the `list` command prints the images required for deploying Karmada under default settings, using the default image repository `registry.k8s.io`.

Reviewer 3: get it, thanks

  2. When the `private-image-registry` is not empty, and the `kube-image-mirror-country` flag is empty, the `list` command prints the addresses of the images required for Karmada under the user-specified image repository link.
  3. When `kube-image-mirror-country` is set to `cn`, the `list` command prints the download links for the images from the default `cn` image address set in `karmadactl`, `registry.cn-hangzhou.aliyuncs.com/google_containers`. For `global` or other addresses not listed in the system, the behavior is similar to that described in point 1, using the default image repository `registry.k8s.io`.
  4. When both `private-image-registry` and `kube-image-mirror-country` flags are not empty, it is recommended to use the `private-image-registry` specified by the user, as described in point 2. Alternatively, it should return an error message. This behavior might require further discussion.

### Implementation of Reading Configuration File

Based on my classification of dozens of commands in the current `karmadactl init` command, I have designed the following data structure for reading:

```go

> go复制代码type InitConfiguration struct {

Author: Some of the above mentioned errors have been repaired, please remind me again if there are more corrections to be made

>     metav1.TypeMeta `json:",inline"`

Reviewer 3: What is `metav1.TypeMeta` used for? Is this an API?

Author: I believe this mimics the configuration file of kubeadm init. I observed that in the currently supported configuration files of kubeadm, there are different version identifiers (such as v1beta3, v1beta4, etc.). In these different versions, the supported configuration fields and the logic for reading the configuration files are different. Therefore, I added metav1.TypeMeta here for future update management.

---
title: Karmadactl Supports Configuration File Method for Production-Grade Installation and Deployment
authors:
  - "@tiansuo114"

> reviewers:

Reviewer 2: reviewer and approver need add more,cc @zhzhuang-zju

  - "@liangyuanpeng"

approvers:
  - "@liangyuanpeng"

creation-date: 2024-07-29
---

# Karmadactl Supports Configuration File Method for Production-Grade Installation and Deployment

## Summary

> Additionally, `karmadactl init` requires the use of multiple container images during deployment, and users currently cannot easily obtain a list of the container images used. Therefore, we plan to add an `images list` subcommand to allow users to print the list of container images used during the `karmadactl init` process to the console.

Reviewer 3: wo can add some summary about supporting configuration files.

Author: Ok, I have modified the section and the following questions
@zhzhuang-zju

After completing the design for reading config files in `karmada init`, we can extend the functionality of the `karmadactl config image list` command to read the required image configuration from the config file by reusing the config file reading method.

### User Stories (Optional)

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### CLI flags changes

This proposal proposes new flags `config`in `karmadactl init` flag set.

| name   | shorthand | default | usage                  |
| ------ | :-------: | ------- | ---------------------- |
| config |     /     | ""      | Karmada init file path |

With these flag, we will:

> - When the user has a karmadactl init file, they can use the `--config` flag to specify that `karmadactl init` should read the configuration from the specified config file. Additionally, external cmd flags can still be used. In the design, if there are overlapping parts between the two, the configuration file configuration items will take higher priority.

Reviewer 3: > the configuration file configuration items will take higher priority.

Is it consistent with `kubeadm`?

Author: Yes, the setup here is the same as `kubeadm init --config`

Reviewer 4: I feel that this content can be highlighted in bold ^-^

- performance：
  1. When the `private-image-registry` and `kube-image-mirror-country` flags are empty, the `list` command prints the images required for deploying Karmada under default settings, using the default image repository `registry.k8s.io`.
  2. When the `private-image-registry` is not empty, and the `kube-image-mirror-country` flag is empty, the `list` command prints the addresses of the images required for Karmada under the user-specified image repository link.
  3. When `kube-image-mirror-country` is set to `cn`, the `list` command prints the download links for the images from the default `cn` image address set in `karmadactl`, `registry.cn-hangzhou.aliyuncs.com/google_containers`. For `global` or other addresses not listed in the system, the behavior is similar to that described in point 1, using the default image repository `registry.k8s.io`.
  4. When both `private-image-registry` and `kube-image-mirror-country` flags are not empty, it is recommended to use the `private-image-registry` specified by the user, as described in point 2. Alternatively, it should return an error message. This behavior might require further discussion

> ### Implementation of Reading Configuration File

Reviewer 3: How about:
```suggestion
### Implementation of Loading Configuration File
```

> Based on my classification of dozens of commands in the current `karmadactl init` command, I have designed the following data structure for reading:

Reviewer 3: ```suggestion
Based on my classification of dozens of flags in the current `karmadactl init` command, I have designed the following data structure:
```

Example configuration file:

```yaml
apiVersion: config.karmada.io/v1beta1
kind: InitConfiguration
generalConfig:
  namespace: "karmada-system"
  kubeConfigPath: "/etc/karmada/kubeconfig"
  kubeImageTag: "v1.21.0"
  context: "karmada-context"
  storageClassesName: "fast"
  privateImageRegistry: "local.registry.com"
  waitComponentReadyTimeout: 120
  port: 32443

certificateConfig:
  externalDNS:
    - "www.karmada.io"
  externalIP:
    - "10.235.1.2"
  validityPeriod: "8760h"
  caCertFile: "/etc/karmada/pki/ca.crt"
  caCertKeyFile: "/etc/karmada/pki/ca.key"
  extraArgs:
    - name: "arg1"
      value: "value1"

etcdConfig:
  local:
    image: "local.registry.com/library/etcd:3.5.13-0"
    initImage: "docker.io/alpine:3.19.1"
    dataDir: "/var/lib/karmada-etcd"
    pvcSize: "5Gi"

>     nodeSelectorLabels: "karmada.io/etcd=true"

Reviewer 2: Could you give an example of multi-label selection and why strings are used instead of arrays?

Author: Is this a discussion about nodeSelectorLabels field? The problems about multi-label case and string and array may be due to my wrong understanding of this field. May this field have options for complex numbers? Please help me further explain

Reviewer 2: check the usage from `kubectl explain sts.spec.template.spec.nodeSelector`

```yaml
nodeSelector:
  kubernetes.io/os: linux
  hello: world
```

Author: Ok, but I notice that in the current karmadactl use case, there is only one EtcdNodeSelectorLabels string field. How do you implement multiple label selection in the current karmadactl init command?

Reviewer 2: > How do you implement multiple label selection in the current karmadactl init command?

Separate multiple label with `.`, like  `EtcdNodeSelectorLabels=kubernetes.io/os=linux,hello=world`, do you want to open a PR for it?

The second option is to only support multiple labels in the configuration file, and eventually eliminate the `EtcdNodeSelectorLabels` on `karmadactl init`

Author: 我觉得可以采用第一种方式,我提出了一个新的issue,我会尽快完成内容
Support multiple label selection ability in karmadactl init

	KubeImageMirrorCountry string   yaml:"kubeImageMirrorCountry"
	ImagePullPolicy        string   yaml:"imagePullPolicy"
	ImagePullSecrets       []string yaml:"imagePullSecrets"
}

// Arg represents a name-value pair argument.
type Arg struct {
	Name  string yaml:"name"
	Value string yaml:"value"
}
```

After reading the logic, the information in the configuration file will be placed into the `InitConfiguration` structure, and then the fields will be filled into the `CommandInitOption` structure used to control the `karmadactl init` process. This way, the deployment of the Karmada control plane can proceed according to the original logic.

For the design of the configuration file, I think it can be designed similarly to the config structure in `kubeadm init`. I will provide a YAML file example below. To provide users with a configuration file, I think we can prepare a default configuration file example and make it a sub-command in `karmadactl config`. When the feature design is complete and the official documentation is updated, an example can also be demonstrated in the official documentation.

Example configuration file:

```yaml

> apiVersion: config.karmada.io/v1beta1

Reviewer 2: ```suggestion
apiVersion: config.karmada.io/v1alpha1
```

Author: have fixed

kind: InitConfiguration
general:
  namespace: "karmada-system"
  kubeConfigPath: "/etc/karmada/kubeconfig"
  kubeImageTag: "v1.21.0"
  context: "karmada-context"
  storageClassesName: "fast"
  privateImageRegistry: "local.registry.com"
  waitComponentReadyTimeout: 120
  port: 32443

certificate:
  externalDNS:
    - "www.karmada.io"
  externalIP:
    - "10.235.1.2"
  validityPeriod: "8760h"
  caCertFile: "/etc/karmada/pki/ca.crt"
  caCertKeyFile: "/etc/karmada/pki/ca.key"
  extraArgs:
    - name: "arg1"
      value: "value1"

etcd:
  local:
    image: "local.registry.com/library/etcd:3.5.13-0"
    initImage: "docker.io/alpine:3.19.1"
    dataDir: "/var/lib/karmada-etcd"
    pvcSize: "5Gi"
    nodeSelectorLabels:
      kubernetes.io/os: linux
      hello: world
    storageMode: "PVC"
    replicas: 3
    extraArgs:
      - name: "etcd-arg1"
        value: "etcd-value1"
  external:
    caPath: "/etc/ssl/certs/ca-certificates.crt"
    certPath: "/path/to/your/certificate.pem"
    keyPath: "/path/to/your/privatekey.pem"
    servers: "https://example.com:8443"
    prefix: "ext-"
    extraArgs:
      - name: "ext-etcd-arg1"
        value: "ext-etcd-value1"

> karmadaControlPlane:
>   apiServer:

Reviewer 2: ```suggestion
apiServer:
```

I think we can same with kubeadm, it's mean the component name as first level, Other ideas welcome

Author: In my initial design, I thought that isolating the configurations required for the Karmada control plane from other configurations might make it more readable. This could potentially be further discussed in the PR

    image: "karmada-apiserver:latest"
    advertiseAddress: "192.168.1.2"
    replicas: 3
    nodePort: 32443
    extraArgs:
      - name: "api-arg1"
        value: "api-value1"
  controllerManager:
    image: "karmada-controller-manager:latest"
    replicas: 3
    extraArgs:
      - name: "cm-arg1"
        value: "cm-value1"
  scheduler:
    image: "karmada-scheduler:latest"
    replicas: 3
    extraArgs:
      - name: "sched-arg1"
        value: "sched-value1"
  webhook:
    image: "karmada-webhook:latest"
    replicas: 3
    extraArgs:
      - name: "webhook-arg1"
        value: "webhook-value1"
  aggregatedAPIServer:
    image: "karmada-aggregated-apiserver:latest"
    replicas: 3
    extraArgs:
      - name: "aggregated-arg1"
        value: "aggregated-value1"
  kubeControllerManager:
    image: "kube-controller-manager:latest"
    replicas: 3
    extraArgs:
      - name: "kcm-arg1"
        value: "kcm-value1"
  dataPath: "/etc/karmada"
  pkiPath: "/etc/karmada/pki"
  crds: "/etc/karmada/crds"
  hostClusterDomain: "cluster.local"

> image:

Reviewer 2: It seems that this section can be placed under `general`

```
general:
  image:
```

Author: I've put the image in general, and I've completed a few instances, so how does it look now

---
title: Karmadactl Supports Configuration File Method for Production-Grade Installation and Deployment
authors:
  - "@tiansuo114"
reviewers:
  - "@liangyuanpeng"
  - "@zhzhuang-zju"
  - "@XiShanYongYe-Chang"

approvers:
  - "@liangyuanpeng"
  - "@zhzhuang-zju"
  - "@XiShanYongYe-Chang"

creation-date: 2024-07-29
---

# Karmadactl Supports Configuration File Method for Production-Grade Installation and Deployment

## Summary

> The `karmadactl init` command now supports reading deployment parameters from a configuration file, simplifying the process of production-grade installation and reducing the complexity of managing numerous command-line flags. This enhancement allows users to streamline their multi-cluster management setups by utilizing predefined configuration files, similar to Kubernetes' `kubeadm`, making it easier to customize and maintain consistent deployments across environments.

Reviewer 3: Using 'now' might not be appropriate since `karmadactl config` is a future feature enhancement.

```suggestion
The `karmadactl init` command is expected to support loading deployment parameters from a configuration file, simplifying the process of production-grade installation and reducing the complexity of managing numerous command-line flags. This enhancement allows users to streamline their multi-cluster management setups by utilizing predefined configuration files, similar to Kubernetes' `kubeadm`, making it easier to customize and maintain consistent deployments across environments.
```

Author: Ok, so far these inaccuracies have been fixed

> In addition to this, we have introduced an `images list` subcommand. This command allows users to easily retrieve and print the list of container images required during the `karmadactl init` process. This feature is particularly useful for users who wish to download the images in advance, reducing deployment time, especially in offline or restricted environments.

Reviewer 3: ```suggestion
In addition to this, we want to introduce an `images list` subcommand. This command allows users to easily retrieve and print the list of container images required during the `karmadactl init` process. This feature is particularly useful for users who wish to download the images in advance, reducing deployment time, especially in offline or restricted environments.
```

- Adds the ability to display the list of images required in karmadactl addon deployment by extending the cmd flag of karmadactl config image

## Proposal

I believe that the flags in `karmada init` can mainly be divided into the following categories:

- Certificate-related Options
- Etcd-related Options
- Karmada Control Plane-related Options
- Kubernetes-related Options
- Networking Options
- Image Options

Therefore, I think the configuration file can be designed to focus on these parts, structuring the YAML accordingly, and incorporating the existing configuration items into them. Additionally, similar to the commonly used `Extra Arg` field in Kubernetes structures, an extra field can be added to store other general fields. When designing the fields, it is advisable to refer to the field names in the kubeadm configuration file as much as possible, to facilitate user familiarity and quick adoption.

After completing the design for reading config files in `karmada init`, we can extend the functionality of the `karmadactl config image list` command to read the required image configuration from the config file by reusing the config file reading method.

> ### User Stories (Optional)

Reviewer 4: Can you provide two user stories to explain the two goals from the perspective of users?

- When the user has a karmadactl init file, they can use the `--config` flag to specify that `karmadactl init` should read the configuration from the specified config file. Additionally, external cmd flags can still be used. In the design, if there are overlapping parts between the two, the configuration file configuration items will take higher priority.

### Adding New Commands

#### karmadactl config image list

- description:Shows information about the images required for Karmada deployment.
- usage scope:Default image and configure the specified image
- image-related flag set:

| name                      | shorthand | default | usage                                                        |
| ------------------------- | --------- | ------- | ------------------------------------------------------------ |
| private-image-registry    | /         | ""      | Private image registry where pull images from. If set, all required images will be downloaded from it, it would be useful in offline installation scenarios. |
| kube-image-mirror-country | /         | ""      | Country code of the kube image registry to be used. For Chinese mainland users, set it to cn |

> - performance：

Reviewer 4: Can you provide some specific examples of different situations to illustrate, so that the performance is more intuitive?

Author: Ok, I've added two user stories and some use cases for both parts of the code, so how does the documentation look now

Author: Ok, I've added two user stories and some use cases for both parts of the code, so how does the documentation look now
@XiShanYongYe-Chang

Reviewer 4: Thank you. It looks much more intuitive now.

  2. When the `private-image-registry` is not empty, and the `kube-image-mirror-country` flag is empty, the `list` command prints the addresses of the images required for Karmada under the user-specified image repository link.
  3. When `kube-image-mirror-country` is set to `cn`, the `list` command prints the download links for the images from the default `cn` image address set in `karmadactl`, `registry.cn-hangzhou.aliyuncs.com/google_containers`. For `global` or other addresses not listed in the system, the behavior is similar to that described in point 1, using the default image repository `registry.k8s.io`.
  4. When both `private-image-registry` and `kube-image-mirror-country` flags are not empty, it is recommended to use the `private-image-registry` specified by the user, as described in point 2. Alternatively, it should return an error message. This behavior might require further discussion

### Implementation of Loading Configuration File

Based on my classification of dozens of flags in the current `karmadactl init` command, I have designed the following data structure:

```go
type InitConfiguration struct {

> 	metav1.TypeMeta json:",inline"

Reviewer 4: What does this variable do?

Reviewer: What do you mean?

In addition to this, we want to introduce an `images list` subcommand. This command allows users to easily retrieve and print the list of container images required during the `karmadactl init` process. This feature is particularly useful for users who wish to download the images in advance, reducing deployment time, especially in offline or restricted environments.

## Motivation

As Karmada is widely used in multi-cluster management, the current `karmadactl init` command has accumulated dozens of command-line parameters. This not only increases the learning curve for users but also reduces the convenience of usage. With the ongoing development of the community, the number and complexity of these parameters are likely to continue growing, further impacting the user experience.

Additionally, during the deployment process of `karmadactl init`, multiple container images are needed. Users may want to download the corresponding images in advance to save deployment time, but currently, there is no easy way for users to obtain the list of required container images.

### Goals

- Provide the ability for `karmadactl` to construct `init` command parameters from a configuration file.

> - Implement the `karmadactl config images list` command to provide a list of container images.

Reviewer: I tend to put the sub-command into the `init` command, as it is intended to be used during the installation process.

Like: `karmadactl init show-images`  // as sub-command.

Reviewer 2: I will support karmadactl config list, because there may be `karmadactl config print`, `kubeadm config validate` ... 

@RainbowMango  WDYT?

Reviewer: The difference between karmadactl and kubeadm is that `kubeadm` is the tool dedicated to installation, but `karmadactl` is a comprehensive tool designed for common maintenance purposes. 
I'm afraid `karmadactl config` would bring ambiguity here.

Reviewer 2: So if we didn't use the config subcommand, we would have multiple subcommands, such as karmadactl init show-images, karmadactl addons show-images, karmadactl init show-images, karmadactl addons show-images and karmadactl init migrate? 

And this limits the extension of the kramdactl config command,  how aboud karmadactl init-config or karmadactl deploy-config?

Reviewer: What's the `migrate` thing?

My feeling about the benefit of having the `show-images` is limited, people can get the image list by document somewhere. That's why I don't think it should be under `karmadactl`.

Another option is, to postpone the show-images, just focus on the init configuration.  and wait for more use cases, like `migrate` mentioned on above.

Author: If obtaining the image list through documentation may not be the most convenient approach for users, as they can quickly convert the default image list into their desired one using commands in the command-line program (e.g., --country cn, etc.), I believe it might be a good idea to change it to a separate command like karmadactl init-images or karmadactl init-config. What are your thoughts on this, mentors?
@RainbowMango  @liangyuanpeng

Reviewer 2: > Another option is, to postpone the show-images, just focus on the init configuration. and wait for more use cases, like mentioned on above.migrate

good point, let's focus on the init configuration and handle show-images separately.

Author: Ok, I moved the changes to the karmadactl image list section into another pr([link](https://github.com/karmada-io/karmada/pull/5623)), which now only cares about the new proposal for the `karmadactl init --config` command

Reviewer 2: cc @RainbowMango for review , thanks

Reviewer: OK.

### Non-Goals
- Add additional commands for karmadactl config, such as karmadactl config --init-config to add the ability to output default initialization config to the command
- Adds the ability to display the list of images required in karmadactl addon deployment by extending the cmd flag of karmadactl config image

## Proposal

I believe that the flags in `karmada init` can mainly be divided into the following categories:

- Certificate-related Options
- Etcd-related Options
- Karmada Control Plane-related Options
- Kubernetes-related Options
- Networking Options

> - Image Options

Reviewer: Not an objection, just a question here.
Do you think the [categories we used at karmada-operator](https://github.com/karmada-io/karmada/blob/master/operator/pkg/apis/operator/v1alpha1/type.go#L80-L115) are also an alternative to categorizing these flags?

Author: I believe they should be quite similar, all using a struct to represent and accept information needed during cluster deployment. Perhaps in future community work, we could reorganize the entire `karmadactl init` command, and redesign the current string-filled struct to align with the struct used in `karmada-operator` (similar to what `kubeadm` did). My understanding might be off, and I would appreciate any corrections if there are any discrepancies.

Reviewer: Yeah, get it. Thanks for the clarification. I will look at the InitConfiguration design.

After the feature is introduced, Ms. Li can execute the `karmadactl config images list` command to retrieve the list of images required for Karmada deployment. She can also select the image repository based on her network conditions (such as her company’s internal image registry). By pre-downloading the images, Ms. Li ensures smooth deployment in an offline environment, reducing deployment time.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### CLI flags changes

This proposal proposes new flags `config`in `karmadactl init` flag set.

| name   | shorthand | default | usage                  |
| ------ | :-------: | ------- | ---------------------- |

> | config |     /     | ""      | Karmada init file path |

Reviewer: ```suggestion
| config |     /     | ""      | Karmada init --config=<file path> |
```
Isn't it?

  
     Command:
  
     ```bash
     karmadactl config images list --private-image-registry=registry.company.com
     ```
  
     Output:
  
     ```
     registry.company.com/karmada-apiserver:v0.9.0
     registry.company.com/karmada-controller-manager:v0.9.0
     ...
     ```
  
  3. For users in China, downloading images from local image sources can significantly speed up the process. When the `kube-image-mirror-country` is set to `cn`, the command will return image addresses from a Chinese mirror (e.g., Alibaba Cloud).
  
     Command:
  
     ```
     karmadactl config images list --kube-image-mirror-country=cn
     ```
  
     Output:
  
     ```
     registry.cn-hangzhou.aliyuncs.com/google_containers/karmada-apiserver:v0.9.0
     registry.cn-hangzhou.aliyuncs.com/google_containers/karmada-controller-manager:v0.9.0
     ...
     ```

### Implementation of Loading Configuration File

Based on my classification of dozens of flags in the current `karmadactl init` command, I have designed the following data structure:

```go

> type InitConfiguration struct {

Reviewer: What do you think of the compatibility issue?
For example, the InitConfigration made for v1.12, is it can be recognized by karmadactl@v1.13?

Reviewer 2: > What do you think of the compatibility issue?

Configuring the version of the API will ensure compatibility and not karmadactl version.
```yaml
apiVersion: config.karmada.io/v1alpha1
```

therefore, there may be two API versions in one karmadactl version, like  `config.karmada.io/v1alpha1` and ` config.karmada.io/v1alpha2` at karmada@v1.x

Reviewer: Then, this is another API just like the karmada operator.  A little bit heavy I feel.

I wonder how Kubeadm deals with its configuration.

Author: In `kubeadm`, multiple versions of the API parsing are maintained (e.g., `v1beta3`, `v1beta4`). When `kubeadm` receives a configuration file, it first analyzes the file version and then selects the appropriate package for parsing based on the version.

Author: > Then, this is another API just like the karmada operator. A little bit heavy I feel.
> 
> I wonder how Kubeadm deals with its configuration.

In `kubeadm`, multiple versions of the API parsing are maintained (e.g., `v1beta3`, `v1beta4`). When `kubeadm` receives a configuration file, it first analyzes the file version and then selects the appropriate package for parsing based on the version.
@RainbowMango

Reviewer 2: > I wonder how Kubeadm deals with its configuration.

This is how kubeadm works

Reviewer 2: Same design here for istioctl: 
```yaml
apiVersion: install.istio. io/v1alpha1
```
I think this is a mainstream design in the kubernetes ecosystem.

Reviewer: Thanks for the clarification, I see the [InitConfiguration](https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta4/) of kubeadm. 

It's not a CRD, just an API-kind configuration, right?
Would anyone be able to show me the code on how kubeadm parses and transforms the configuration between versions?

Reviewer 2: >  It's not a CRD, just an API-kind configuration, right?

exactly.

 > how kubeadm parses and transforms the configuration between versions?
 
in kubeadm, you need to use command `kubeadm config migrate` to migrate your configurate file, and only working for `kubeadm.k8s.io/v1beta4`,  prior to this, you had to manually update the configuration file to upgrade the API version.

@tiansuo114 please help to check the code of kubeadm,thank you!

Author: In kubeadm, since it supports importing multiple configuration files at the same time, when reading the configuration files, it first parses them into the following structure:

```go
// DocumentMap is a convenient way to describe a map between a YAML document and its GVK type

// +k8s:deepcopy-gen=false

type DocumentMap map[schema.GroupVersionKind][]byte

// GroupVersionKind unambiguously identifies a kind.  It doesn't anonymously include GroupVersion

// to avoid automatic coercion.  It doesn't use a GroupVersion to avoid custom marshalling

type GroupVersionKind struct {

	Group   string

	Version string

	Kind    string

}

```

Then, in the documentMapToInitConfiguration method that parses the configuration files, a specific version decoder is designated through the following code to handle version control:

```go
	for _, gvk := range gvks {
		fileContent := gvkmap[gvk]

		// first, check if this GVK is supported and possibly not deprecated
		if err := validateSupportedVersion(gvk, allowDeprecated, allowExperimental); err != nil {
			return nil, err
		}

		// verify the validity of the YAML
		if err := strict.VerifyUnmarshalStrict([]*runtime.Scheme{kubeadmscheme.Scheme, componentconfigs.Scheme}, gvk, fileContent); err != nil {
			if !strictErrors {
				klog.Warning(err.Error())
			} else {
				return nil, err
			}
		}

		if kubeadmutil.GroupVersionKindsHasInitConfiguration(gvk) {
			// Set initcfg to an empty struct value the deserializer will populate
			initcfg = &kubeadmapi.InitConfiguration{}
			// Decode the bytes into the internal struct. Under the hood, the bytes will be unmarshalled into the
			// right external version, defaulted, and converted into the internal version.
			if err := runtime.DecodeInto(kubeadmscheme.Codecs.UniversalDecoder(), fileContent, initcfg); err != nil {
				return nil, err
			}
			continue
		}
		if kubeadmutil.GroupVersionKindsHasClusterConfiguration(gvk) {
			// Set clustercfg to an empty struct value the deserializer will populate
			clustercfg = &kubeadmapi.ClusterConfiguration{}
			// Decode the bytes into the internal struct. Under the hood, the bytes will be unmarshalled into the
			// right external version, defaulted, and converted into the internal version.
			if err := runtime.DecodeInto(kubeadmscheme.Codecs.UniversalDecoder(), fileContent, clustercfg); err != nil {
				return nil, err
			}
			continue
		}

		// If the group is neither a kubeadm core type or of a supported component config group, we dump a warning about it being ignored
		if !componentconfigs.Scheme.IsGroupRegistered(gvk.Group) {
			klog.Warningf("[config] WARNING: Ignored YAML document with GroupVersionKind %v\n", gvk)
		}
	}
```

I believe the control should be based on the above logic. If I am mistaken, please let me know, and I will promptly correct it.
@liangyuanpeng  @RainbowMango

Reviewer 2: So kubeadm only handles the corresponding API version in the configuration file and does not automatically use the v2 version to handle v1 API logic, right? @tiansuo114

Reviewer: OK. It makes sense to me. Thanks.

Author: > So kubeadm only handles the corresponding API version in the configuration file and does not automatically use the v2 version to handle v1 API logic, right? 那么kubeadm只处理配置文件中对应的API版本，并不会自动使用v2版本来处理v1 API逻辑，对吗？@tiansuo114

Currently, it should work like this: it will handle the corresponding v1beta3 and v1beta4 configuration files and map them to the appropriate decoders for each version (here, the v1beta3 decoder will not be used to handle v1beta4 configuration files, and vice versa). As for older versions of the configuration files, users will be prompted to upgrade. However, I am not sure about the v2 version. At present, the latest version of the kubeadm init config is v1beta4, and there doesn't seem to be a v2 version yet.

@liangyuanpeng

Reviewer 2: >  I am not sure about the v2 version. At present, the latest version of the kubeadm init config is v1beta4, and there doesn't seem to be a v2 version yet.

yep, I'm just giving random examples

	metav1.TypeMeta json:",inline"

	// GeneralConfig contains the common configuration at initialization.
	GeneralConfig GeneralConfig yaml:"general"

	// CertificateConfig contains configurations related to certificates.
	CertificateConfig CertificateConfig yaml:"certificate"

	// EtcdConfig contains configuration for etcd.
	EtcdConfig EtcdConfig yaml:"etcd"

	// KarmadaControlPlaneConfig contains control plane configurations.
	KarmadaControlPlaneConfig KarmadaControlPlaneConfig yaml:"karmadaControlPlane"
}

// GeneralConfig contains general configuration parameters.
type GeneralConfig struct {
	Namespace                 string yaml:"namespace"
	KubeConfigPath            string yaml:"kubeConfigPath"

> 	KubeImageTag              string yaml:"kubeImageTag"

Author: I noticed the cmd in the original init command Flag, use KubeImageTag or KarmadaAPIServerImage and KubeControllerManagerImage respectively control the cluster service using the function of the image, I think his appearance here is not redundant? This option may be useful in situations where users only want to specify the version of the image rather than the default image address

	KubeContextName           string yaml:"context"

> 	KubeStorageClassesName    string yaml:"storageClassesName"

Reviewer: OK. By the way, can you add some comments to the structure? If there is a corresponding flag, please also put it in comments.

> 	KubeStorageClassesName    string yaml:"storageClassesName"

Author: Ok, so far, I have only added explanatory comments before each general structure, but I think the information in the corresponding fields of each structure should confirm their purpose. In your opinion, where should you add comments about the corresponding flag? Do you mean the corresponding flag of which configuration items?

	PrivateImageRegistry      string yaml:"privateImageRegistry"
	WaitComponentReadyTimeout int    yaml:"waitComponentReadyTimeout"

> 	Port                      int    yaml:"port"

Reviewer: The `Port` is dedicated to karmada-apiserver. Should categorize into APIServerConfig.

Author: Yes, the design here is a redundant error, at the beginning I will. Its design is used for filling CommandInitOption KarmadaAPIServerNodePort, but behind the changes

> 	StorageClassesName        string yaml:"storageClassesName"

Reviewer: The KubeStorageClassesName is dedicated to etcd, right? It shouldn't be present in GeneralConfig I guess.

Author: KubeStorageClassesName is indeed redundant, I will change the document immediately, the problem here is the same as described above
